### PR TITLE
Update to aws_cryptosdk_enc_ctx_size proof

### PR DESCRIPTION
**Pull request by tegansb**
*Description of changes:*

Added non-NULL assumptions to harness of proof for enc_ctx_size. Updated the generator for the hash with assumptions that the key and value generated are non-NULL. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c [hunk 1]
```diff
@@ -31,22 +31,27 @@ void hash_iterator_generator(struct aws_hash_iter *new_iter, const struct aws_ha
...
+        __CPROVER_assume(new_iter->element.key != NULL);
         new_iter->element.value = malloc(sizeof(struct aws_string));
+        __CPROVER_assume(new_iter->element.value != NULL);
     }
 }
 
 void aws_cryptosdk_enc_ctx_size_harness() {
+    /* Nondet Input */
     struct aws_hash_table map;
+    size_t size;
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:44]: It would be clearer to malloc these, then assume nonnull.

**Review by danielsn — CHANGES_REQUESTED** [review]

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c (hunk 1 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c]: Updated!

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c [hunk 2]
```diff
@@ -31,22 +31,29 @@ void hash_iterator_generator(struct aws_hash_iter *new_iter, const struct aws_ha
 void hash_iterator_generator2(struct aws_hash_iter *new_iter, const struct aws_hash_iter *old_iter) {
     (void)old_iter;
     if (new_iter->status == AWS_HASH_ITER_STATUS_READY_FOR_USE) {
-        new_iter->element.key   = malloc(sizeof(struct aws_string));
+        new_iter->element.key = malloc(sizeof(struct aws_string));
+        __CPROVER_assume(new_iter->element.key != NULL);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:35]: Why not assume an invariant here (i.e., `aws_string_is_valid()`)? What is the corresponding precondition in the function?

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c [hunk 3]
```diff
@@ -31,22 +31,29 @@ void hash_iterator_generator(struct aws_hash_iter *new_iter, const struct aws_ha
...
         new_iter->element.value = malloc(sizeof(struct aws_string));
+        __CPROVER_assume(new_iter->element.value != NULL);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:37]: Why not assume an invariant here (i.e., `aws_string_is_valid()`)? What is the corresponding precondition in the function?

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c [hunk 4]
```diff
@@ -31,22 +31,29 @@ void hash_iterator_generator(struct aws_hash_iter *new_iter, const struct aws_ha
...
 }
 
 void aws_cryptosdk_enc_ctx_size_harness() {
-    struct aws_hash_table map;
-    ensure_allocated_hash_table(&map, MAX_TABLE_SIZE);
-    __CPROVER_assume(aws_hash_table_is_valid(&map));
-    ensure_hash_table_has_valid_destroy_functions(&map);
+    /* Nondet Input */
+    struct aws_hash_table *map = can_fail_malloc(sizeof(*map));
+    size_t *size = can_fail_malloc(sizeof(size));
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:44]: Why are you using `can_fail_malloc` here?

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c [hunk 5]
```diff
@@ -31,22 +31,29 @@ void hash_iterator_generator(struct aws_hash_iter *new_iter, const struct aws_ha
...
 
+    /* Assumptions */
+    __CPROVER_assume(map != NULL);
+    ensure_allocated_hash_table(map, MAX_TABLE_SIZE);
+    __CPROVER_assume(aws_hash_table_is_valid(map));
+    ensure_hash_table_has_valid_destroy_functions(map);
     size_t empty_slot_idx;
-    __CPROVER_assume(aws_hash_table_has_an_empty_slot(&map, &empty_slot_idx));
+    __CPROVER_assume(aws_hash_table_has_an_empty_slot(map, &empty_slot_idx));
+    __CPROVER_assume(size != NULL);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:53]: Where is the corresponding precondition in the actual function?

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c [hunk 6]
```diff
@@ -31,22 +31,29 @@ void hash_iterator_generator(struct aws_hash_iter *new_iter, const struct aws_ha
...
 
-    size_t size;
-    // Precondition: both params are backed by valid memory
-    int rval = aws_cryptosdk_enc_ctx_size(&size, &map);
-    assert(aws_hash_table_is_valid(&map));
+    /* Operation under verification */
+    int rval = aws_cryptosdk_enc_ctx_size(size, map);
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:56]: Add an extra empty line after this one.

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c (hunk 2 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:35]: Updated to use the invariant and precondition is added to the function.

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c (hunk 3 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:37]: Same as above.

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c (hunk 4 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:44]: Updated!

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c (hunk 6 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:56]: Done!

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c (hunk 5 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:53]: Yeah that's a good point. I've looked at other proofs that use the aws_hash_table_has_an_empty_slot function and none of them have a corresponding precondition in the harness. I do think this should be added, but maybe as a separate PR. I can't think of a simple, easy fix.

Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:53]: Having an empty slot is implied by this line in the `hash_table_is_valid` function. Assuming that the hash-table is correctly implemented, such that all non-filled slots are marked empty, then having fewer filled slots than available slots ensures this property.
https://github.com/awslabs/aws-c-common/blob/master/source/hash_table.c#L1040
But there is not an explicit check that this is actually true. The difficulty is that this requires checking for the existance of such an actual slot, which requires a potentially expensive O(n) loop, whereas the rest of the check of O(1) time.

File: source/enc_ctx.c [hunk 7]
```diff
@@ -56,8 +56,10 @@ int aws_cryptosdk_enc_ctx_size(size_t *size, const struct aws_hash_table *enc_ct
     size_t serialized_len = 2;  // First two bytes are the number of k-v pairs
     for (struct aws_hash_iter iter = aws_hash_iter_begin(enc_ctx); !aws_hash_iter_done(&iter);
          aws_hash_iter_next(&iter)) {
-        const struct aws_string *key   = iter.element.key;
+        const struct aws_string *key = iter.element.key;
+        AWS_PRECONDITION(aws_string_is_valid(key));
```
Comment (danielsn) [inline source/enc_ctx.c:60]: This isn't really a precondition at this point - maybe just make it a regular assert?

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c (hunk 5 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:53]: Sounds good and we do have the hash_table_is_valid precondition on the function.

File: source/enc_ctx.c (hunk 7 above)
Comment (tegansb) [reply source/enc_ctx.c:60]: Updated!

**Review by danielsn — APPROVED** [review]

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c [hunk 8]
```diff
@@ -31,22 +31,30 @@ void hash_iterator_generator(struct aws_hash_iter *new_iter, const struct aws_ha
...
 }
 
 void aws_cryptosdk_enc_ctx_size_harness() {
-    struct aws_hash_table map;
-    ensure_allocated_hash_table(&map, MAX_TABLE_SIZE);
-    __CPROVER_assume(aws_hash_table_is_valid(&map));
-    ensure_hash_table_has_valid_destroy_functions(&map);
+    /* Nondet Input */
+    struct aws_hash_table *map = malloc(sizeof(*map));
+    size_t *size               = malloc(sizeof(size));
```
Comment (feliperodri) [inline .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:44]: malloc(sizeof(*size));

**Review by feliperodri — APPROVED** [review]
@tegansb one final correction, but overall LGTM.

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c (hunk 8 above)
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c:44]: nice catch.

Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_size/aws_cryptosdk_enc_ctx_size_harness.c]: Yes, thank you! Fixed.

**Review by alex-chew — APPROVED** [review]
LGTM

**Pull request merged**